// @superfluid-finance/ethereum-contracts/build/typechain/factories/contracts/mocks/CFAHookMocks.sol/BaseCFAHookMock__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCFAHookMock__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "NOT_CFA",
        type: "error",
    },
    {
        inputs: [],
        name: "NOT_OWNER",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "contract ISuperfluidToken",
                name: "token",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "address",
                        name: "sender",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "receiver",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "flowOperator",
                        type: "address",
                    },
                    {
                        internalType: "int96",
                        name: "flowRate",
                        type: "int96",
                    },
                ],
                internalType: "struct IConstantFlowAgreementHook.CFAHookParams",
                name: "newFlowData",
                type: "tuple",
            },
        ],
        name: "onCreate",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract ISuperfluidToken",
                name: "token",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "address",
                        name: "sender",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "receiver",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "flowOperator",
                        type: "address",
                    },
                    {
                        internalType: "int96",
                        name: "flowRate",
                        type: "int96",
                    },
                ],
                internalType: "struct IConstantFlowAgreementHook.CFAHookParams",
                name: "newFlowData",
                type: "tuple",
            },
            {
                internalType: "int96",
                name: "oldFlowRate",
                type: "int96",
            },
        ],
        name: "onDelete",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract ISuperfluidToken",
                name: "token",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "address",
                        name: "sender",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "receiver",
                        type: "address",
                    },
                    {
                        internalType: "address",
                        name: "flowOperator",
                        type: "address",
                    },
                    {
                        internalType: "int96",
                        name: "flowRate",
                        type: "int96",
                    },
                ],
                internalType: "struct IConstantFlowAgreementHook.CFAHookParams",
                name: "newFlowData",
                type: "tuple",
            },
            {
                internalType: "int96",
                name: "oldFlowRate",
                type: "int96",
            },
        ],
        name: "onUpdate",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "contract ConstantFlowAgreementV1",
                name: "_cfaV1",
                type: "address",
            },
        ],
        name: "setCFA",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
class BaseCFAHookMock__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.BaseCFAHookMock__factory = BaseCFAHookMock__factory;
BaseCFAHookMock__factory.abi = _abi;
